microsoft/rushstack
# [rush-lib] fix: update shrinkwrap when globalPackageExtensions has been changed

## Summary



Fixes #4901

## Details



Add one more condition in WorkspaceInstallManager to compare the values of:

- pnpm-config.json's `globalPackageExtensions` checksum
- shrinkwrap's `packageExtensionsChecksum`

To calculate the checksum, I added the same implementation as per what's used in pnpm ([ref](https://github.com/pnpm/pnpm/blob/ba9409ffcef0c36dc1b167d770a023c87444822d/pkg-manager/core/src/install/index.ts#L821-L824)). The code snippet used in pnpm is as such:

```ts
const packageExtensionsChecksum = isEmpty(opts.packageExtensions ?? {}) ? undefined : createObjectChecksum(opts.packageExtensions!)


export function createObjectChecksum (obj: Record<string, unknown>) {
  const s = JSON.stringify(sortKeys(obj, { deep: true }))
  return crypto.createHash('md5').update(s).digest('hex')
}
```

Since it depends on `sortKeys` (which pnpm uses [sort-keys@4.2.0](https://github.com/sindresorhus/sort-keys/tree/v4.2.0) npm package), I asked in [zulip](https://rushstack.zulipchat.com/#narrow/stream/279883-contributor-helpline/topic/Adding.20external.20dependencies.20or.20vendoring.20them.3F) if it's okay to add a new external dependency or not. The reply I received is to implement it in `node-core-library`, so I added: `Sort.sortKeys` with the tests imported and modified from [sort-keys@4.2.0](https://github.com/sindresorhus/sort-keys/blob/v4.2.0/test.js). 


## How it was tested




1. Run `testrush update` in a Rush repo
2. Change `globalPackageExtensions`
3. Run `testrush update` and expect that it updates the lockfile

```
> testrush update

Rush Multi-Project Build Tool 5.133.3 (unmanaged) - https://rushjs.io
Node.js version is 18.20.4 (LTS)


Starting "rush update"

Checking Git policy for this repository.


Found files in the "common/git-hooks" folder.
Successfully installed these Git hook scripts: commit-msg, owners.json, post-checkout, pre-commit, pre-push

Trying to acquire lock for pnpm-8.15.8
Acquired lock for pnpm-8.15.8
Found pnpm version 8.15.8 in /path/to/user/.rush/node-v18.20.4/pnpm-8.15.8

Symlinking "/path/to/repo/common/temp/pnpm-local"
  --> "/path/to/user/.rush/node-v18.20.4/pnpm-8.15.8"
Transforming /path/to/repo/common/config/rush/.npmrc
  --> "/path/to/repo/common/temp/.npmrc"

Updating workspace files in /path/to/repo/common/temp
Copying "/path/to/repo/common/config/rush/pnpm-lock.yaml"
  --> "/path/to/repo/common/temp/pnpm-lock.yaml"
Copying "/path/to/repo/common/config/rush/pnpm-lock.yaml"
  --> "/path/to/repo/common/temp/pnpm-lock-preinstall.yaml"

The shrinkwrap file (pnpm-lock.yaml) contains the following issues:
  The package extension hash doesn't match the current shrinkwrap.


Checking installation in "/path/to/repo/common/temp"

Running "pnpm install" in /path/to/repo/common/temp
```

## Impacted documentation



Updated the JSON schema of `node-core-library`
